File: Data Structures & Algorithms/asteroid-collision/submission-7.py
class Solution:
    def asteroidCollision(self, asteroids: List[int]) -> List[int]:
        stack = []

        for a in asteroids:
            isAlive = True

            while stack and stack[-1] > 0 and a < 0:
                if stack[-1] > -a:
                    isAlive = False
                    break
                elif stack[-1] < -a:
                    stack.pop()
                else:
                    stack.pop()
                    isAlive = False
                    break

            if isAlive == True:
                stack.append(a)

        return stack

File: Data Structures & Algorithms/decode-string/submission-7.java
class Solution {
    public String decodeString(String s) {
        Stack<String> stack = new Stack<>();

        for (char ch: s.toCharArray()) {
            if (ch != ']') {
                stack.push(String.valueOf(ch));
            } else {
                String curr = "";
                while (!stack.peek().equals("[")) {
                    curr = stack.pop() + curr;
                }
                stack.pop();

                String num = "";
                while (!stack.isEmpty() && Character.isDigit(stack.peek().charAt(0))) {
                    num = stack.pop() + num;
                }
                int repeat = Integer.parseInt(num);
                StringBuilder sb = new StringBuilder();
                for (int i=0; i<repeat; i++) {
                    sb.append(curr);
                }
                stack.push(sb.toString());
            }
        }
        StringBuilder ans = new StringBuilder();
        for (String str: stack) {
            ans.append(str);
        }
        return ans.toString();
    }
}